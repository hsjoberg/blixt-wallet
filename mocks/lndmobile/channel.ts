import { lnrpc } from "../../proto/proto";

// export const openChannel = async (pubkey: string, amount: number): Promise<lnrpc.ChannelPoint> => {
//   const response = await sendCommand<lnrpc.IOpenChannelRequest, lnrpc.OpenChannelRequest, lnrpc.ChannelPoint>({
//     request: lnrpc.OpenChannelRequest,
//     response: lnrpc.ChannelPoint,
//     method: "OpenChannelSync",
//     options: {
//       nodePubkeyString: pubkey,
//       localFundingAmount: amount,
//       minConfs: 1,
//     },
//   });
//
//   return response;
// };
//
// export const closeChannel = async (fundingTxId: string, outputIndex: number): Promise<string> => {
//   const response = await sendStreamCommand<lnrpc.ICloseChannelRequest, lnrpc.CloseChannelRequest>({
//     request: lnrpc.CloseChannelRequest,
//     method: "CloseChannel",
//     options: {
//       channelPoint: {
//         fundingTxidStr: fundingTxId,
//         outputIndex,
//       },
//     },
//   });
//   return response;
// };

export const pendingChannels = jest.fn(async (): Promise<lnrpc.PendingChannelsResponse> => {
  const response = lnrpc.PendingChannelsResponse.create({
    pendingClosingChannels: [],
    pendingForceClosingChannels: [],
    pendingOpenChannels: [],
    waitingCloseChannels: [],
    totalLimboBalance: 0,
  });
  return response;
});

export const listChannels = jest.fn(async (): Promise<lnrpc.ListChannelsResponse> => {
  const response = lnrpc.ListChannelsResponse.create({
    channels: [], // TODO
  });
  return response;
});

export const channelBalance = jest.fn(async (): Promise<lnrpc.ChannelBalanceResponse> => {
  const response = lnrpc.ChannelBalanceResponse.create({
    balance: 0, // TODO
    pendingOpenBalance: 0,
  });
  return response;
});

export const subscribeChannelEvents = jest.fn(async (address: string, sat: number): Promise<string> => {
  return ""; // TODO(hsjoberg)
});

export const exportAllChannelBackups = async (): Promise<lnrpc.ChanBackupSnapshot> => {
  const response = lnrpc.ChanBackupSnapshot.create({
    multiChanBackup: {
      // chanPoints
      multiChanBackup: new Uint8Array([1,2,3,4,5,6,7,8,9,10]),
    },
  });
  return response;
};