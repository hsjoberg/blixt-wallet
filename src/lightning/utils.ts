import deepMapKeys from "deep-map-keys";

export const fixGrpcJsonResponse = <T>(response: { [keys: string]: any })  =>
  deepMapKeys<T>(response, (key) => {
    if (key[key.length - 1] === "_") {
      return key.slice(0, -1);
    }
    return "$" + key;
});
