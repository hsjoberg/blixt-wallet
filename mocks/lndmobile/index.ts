import { lnrpc, routerrpc } from "../../proto/lightning";
import { decodeStreamResult, timeout } from "./utils";
import { DeviceEventEmitter } from "react-native";
import * as base64 from "base64-js";
import payReq from "bolt11";
import Long from "long";
import { IAddInvoiceBlixtLspArgs } from "../../src/lndmobile";
import { bytesToHexString } from "../../src/utils";

export enum ELndMobileStatusCodes {
  STATUS_SERVICE_BOUND = 1,
  STATUS_PROCESS_STARTED = 2,
  STATUS_WALLET_UNLOCKED = 4,
}

let LndMobileStatus = 0;

export const initialize = jest.fn(async (): Promise<{ data: string } | number> => {
  await timeout(10);
  LndMobileStatus += ELndMobileStatusCodes.STATUS_SERVICE_BOUND; // TODO figure out bitmasking...
  return { data: "" };
});

export const checkStatus = jest.fn(async (): Promise<number> => {
  return LndMobileStatus;
});

export const writeConfig = jest.fn(async () => {
  return "File written:";
});

export const writeConfigFile = jest.fn(async () => {
  return "File written:";
});

export const subscribeStateEmitter = (data: Uint8Array) => DeviceEventEmitter.emit("SubscribeState", { data: base64.fromByteArray(data) });
export const subscribeState = async () => {
  setTimeout(async () => {
    subscribeStateEmitter(
      lnrpc.SubscribeStateResponse.encode({
        state: lnrpc.WalletState.LOCKED,
      }).finish()
    );
  }, 10)
}

export const decodeState = (data: string): lnrpc.SubscribeStateResponse => {
  return decodeStreamResult<lnrpc.SubscribeStateResponse>({
    response: lnrpc.SubscribeStateResponse,
    base64Result: data,
  });
};

export const startLnd = jest.fn(async (): Promise<string> => {
  await timeout(100);
  return "started";
});

// export const connectPeer = async (pubkey: string, host: string): Promise<lnrpc.ConnectPeerResponse> => {
//   return await sendCommand<lnrpc.IConnectPeerRequest, lnrpc.ConnectPeerRequest, lnrpc.ConnectPeerResponse>({
//     request: lnrpc.ConnectPeerRequest,
//     response: lnrpc.ConnectPeerResponse,
//     method: "ConnectPeer",
//     options: {
//       addr: lnrpc.LightningAddress.create({
//         host,
//         pubkey,
//       }),
//     },
//   });
// };
//
// export const getNodeInfo = async (pubKey: string): Promise<lnrpc.NodeInfo> => {
//   const response = await sendCommand<lnrpc.INodeInfoRequest, lnrpc.NodeInfoRequest, lnrpc.NodeInfo>({
//     request: lnrpc.NodeInfoRequest,
//     response: lnrpc.NodeInfo,
//     method: "GetNodeInfo",
//     options: {
//       pubKey,
//     },
//   });
//   return response;
// };

export const getInfoResponse = lnrpc.GetInfoResponse.create({
  uris: [],
  chains: [{
    chain: "bitcoin",
    network: "testnet",
  }],
  identityPubkey: "02b5380da0919e32b13c1a21c1c85000eed0ba9a9309fc6849d72230d43088ae1d",
  alias: "02b5380da0919e32b13c",
  numPeers: 3,
  blockHeight: 1572555,
  blockHash: "000000000000006cb43faa5c615e45419f48e9d94d77c1bab8a28018cf2db6ef",
  syncedToChain: false,
  syncedToGraph: false,
  testnet: true,
  bestHeaderTimestamp: Long.fromNumber(1564940232),
  version: "0.7.1-beta commit=v0.7.1-beta-rc1-10-g3760f29f5e758b2865b756604333ca22cf23e90b",
  features: {},
});
export const getInfo = jest.fn()
  .mockImplementationOnce(async () => getInfoResponse)
  .mockImplementation(async () =>  ({ ...getInfoResponse, syncedToChain: true, syncedToGraph: true }));

export const sendPaymentSync = async (paymentRequest: string): Promise<lnrpc.SendResponse> => {
  const response = lnrpc.SendResponse.create({
    paymentHash: new Uint8Array([1,2,3,4]),
    paymentRoute: {
      totalAmt: Long.fromNumber(1000),
      totalAmtMsat: Long.fromNumber(1000000),
      totalFees: Long.fromNumber(1),
      totalFeesMsat: Long.fromNumber(1000),
      hops: [],
    },
  });
  return response;
};

export const sendPaymentV2Sync = async (paymentRequest: string): Promise<lnrpc.Payment> => {
  await timeout(600);

  const paymentHash = new Uint8Array([0,1,2,3]);
  const paymentPreimage = new Uint8Array([0,1,2,3]);

  const response = lnrpc.Payment.create({
    paymentHash: bytesToHexString(paymentHash),
    paymentPreimage: bytesToHexString(paymentPreimage),

    status: lnrpc.Payment.PaymentStatus.SUCCEEDED,
    fee: Long.fromValue(1),
    feeMsat: Long.fromValue(1000),
    htlcs: [{
      route: {
        hops: [{
          chanId: Long.fromValue(1),
          chanCapacity: Long.fromValue(10000),
          amtToForward: Long.fromValue(100),
          amtToForwardMsat: Long.fromValue(100000),
          fee: Long.fromValue(1),
          feeMsat: Long.fromValue(1000),
          expiry: 3600,
          pubKey: "abc",
        }],
      },
    }],
  });
  return response;
};

export const addInvoice = async (amount: number, memo: string, expiry: number = 3600): Promise<lnrpc.AddInvoiceResponse> => {
  try {
    const unixTimestamp = Math.floor(Date.now() / 1000);
    const encoded = payReq.encode({
      coinType: "testnet",
      satoshis: amount,
      timestamp: unixTimestamp,
      timeExpireDate: expiry,
      tags: [{
        tagName: "payment_hash",
        data: "0001020304050607080900010203040506070809000102030405060708090102",
      }, {
        tagName: "description",
        data: memo,
      }],
    });

    const privateKeyHex = "e126f68f7eafcc8b74f54d269fe206be715000f94dac067d1c04a8ca3b2db734";
    const signed = payReq.sign(encoded, privateKeyHex);

    const response = lnrpc.AddInvoiceResponse.create({
      paymentRequest: signed!.paymentRequest,
      addIndex: Long.fromNumber(0),
      rHash: new Uint8Array([1, 2, 3]), // TODO
    });

    setTimeout(() => {
      const description = signed!.tags.find((tag) => tag.tagName === "d");
      const cltvExpiry = signed!.tags.find((tag) => tag.tagName === "min_final_cltv_expiry");
      const invoice = lnrpc.Invoice.create({
        paymentRequest: signed!.paymentRequest,
        private: false,
        memo: description && (description.data as string),
        addIndex: Long.fromNumber(0), // TODO
        amtPaid: Long.fromNumber(100),
        rHash: new Uint8Array([1, 2, 3]), // TODO
        value: Long.fromNumber(signed!.satoshis!),
        amtPaidMsat: Long.fromNumber(signed!.satoshis!).mul(1000),
        amtPaidSat: Long.fromNumber(signed!.satoshis!),
        cltvExpiry: cltvExpiry && (Long.fromNumber(cltvExpiry.data as number)),
        creationDate: Long.fromNumber(signed!.timestamp!),
        expiry: Long.fromNumber(expiry),
        rPreimage: new Uint8Array([1, 2, 3, 4]), // TODO
        settled: false,
        state: lnrpc.Invoice.InvoiceState.OPEN,
      });
      DeviceEventEmitter.emit("SubscribeInvoices", {
        data: base64.fromByteArray(lnrpc.Invoice.encode(invoice).finish()),
      });
    }, 5);

    return response;
  } catch (e) {
    console.log(e);
    throw e;
  }
};

export const addInvoiceBlixtLsp = ({amount, memo, expiry = 600, servicePubkey, chanId, cltvExpiryDelta, feeBaseMsat, feeProportionalMillionths, preimage}: IAddInvoiceBlixtLspArgs) => {
  return addInvoice(amount, memo, expiry);
}

//
// export const lookupInvoice = async (rHash: string): Promise<lnrpc.Invoice> => {
//   const response = await sendCommand<lnrpc.IPaymentHash, lnrpc.PaymentHash, lnrpc.Invoice>({
//     request: lnrpc.PaymentHash,
//     response: lnrpc.Invoice,
//     method: "LookupInvoice",
//     options: {
//       rHashStr: rHash,
//     },
//   });
//   return response;
// };

export const queryRoutes = async (pubKey: string, amount?: Long, routeHints?: lnrpc.IRouteHint[]): Promise<lnrpc.QueryRoutesResponse> => {
  return lnrpc.QueryRoutesResponse.create({
    routes: [],
    successProb: 0.5,
  });
}

// TODO test
export const decodePayReq = async (bolt11: string): Promise<lnrpc.PayReq> => {
  const payreq = payReq.decode(bolt11); // todo check second argument
  const description = payreq.tags.find((tag) => tag.tagName === "d");
  const response = lnrpc.PayReq.create({
    cltvExpiry: Long.fromNumber(144), // TODO
    description: description && (description.data as string),
    destination: payreq.payeeNodeKey,
    expiry: Long.fromNumber(payreq.timeExpireDate! - payreq.timeExpireDate!),
    numSatoshis: Long.fromNumber(payreq.satoshis || 0),
    timestamp: Long.fromNumber(payreq.timeExpireDate || 0),
    paymentHash: Math.floor(Math.random() * 1000000).toString(),
  });
  return response;
};

/**
 * @throws
 */
 export const getRecoveryInfo = async (): Promise<lnrpc.GetRecoveryInfoResponse> => {
  const response = lnrpc.GetRecoveryInfoResponse.create({
    progress: 1,
    recoveryFinished: false,
    recoveryMode: false,
  })
  return response;
};

/**
 * @throws
 */
 export const listUnspent = async (): Promise<lnrpc.ListUnspentResponse> => {
  const response = lnrpc.ListUnspentResponse.create({
    utxos: []
  })
  return response;
};

/**
 * @throws
 */
 export const resetMissionControl = async (): Promise<routerrpc.ResetMissionControlResponse> => {
  const response = routerrpc.ResetMissionControlResponse.create({})
  return response;
};

// export type IReadLndLogResponse = string[];
//
// export const readLndLog = async (): Promise<IReadLndLogResponse> => {
//   return [""];
// };
