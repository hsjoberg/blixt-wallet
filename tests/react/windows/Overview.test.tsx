import React from "react";
import { render, toJSON, wait, within, act } from "@testing-library/react-native";
import { StoreProvider } from "easy-peasy";
import Long from "long";
import * as base64 from "base64-js";

import Overview from "../../../src/windows/Overview";
import { setupStore, createNavigationContainer, initCommonStore } from "../../utils";
import { createStackNavigator } from "@react-navigation/stack";
import { DeviceEventEmitter } from "react-native";
import { lnrpc } from "../../../proto/proto";
import { hexToUint8Array } from "../../../src/utils";
import { channelBalance } from "../../../mocks/lndmobile/channel";

it("renders correctly", () => {
  const AppContainer = createNavigationContainer(Overview, "Overview");

  const store = setupStore();
  store.getActions().channel.setBalance(Long.fromNumber(123));

  const { container } = render(
    <StoreProvider store={store}>
      {AppContainer}
    </StoreProvider>
  );
  expect(toJSON(container)).toMatchSnapshot();
});

it("expect balance to update when paying an invoice", async () => {
  let balance = 100000;

  channelBalance.mockImplementation(() => {
    const response = lnrpc.ChannelBalanceResponse.create({
      balance: Long.fromValue(balance),
      pendingOpenBalance: Long.fromValue(0),
    });
    return response;
  });

  const AppContainer = createNavigationContainer(Overview, "Overview");
  const store = await initCommonStore(true);
  // store.getActions().channel.setBalance(Long.fromValue(100000));
  const addInvoiceResponse = await store.getActions().receive.addInvoice({
    sat: 100,
    description: "Receiving 100 sats",
  });
  await wait(() => expect(store.getState().transaction.transactions).toHaveLength(1));

  const { queryByTestId, getByTestId } = render(
    <StoreProvider store={store}>
      {AppContainer}
    </StoreProvider>
  );

  const txList = getByTestId("TX_LIST");
  const tx = store.getState().transaction.transactions[0];
  expect(tx.status).toBe("OPEN");
  expect(within(txList).queryAllByText("Open")).not.toBeNull();

  const bigBalanceHeader = getByTestId("BIG_BALANCE_HEADER");
  expect(bigBalanceHeader.children[0]).toContain("0.001");

  // Pay invoice
  const invoice = lnrpc.Invoice.create({
    amtPaid: Long.fromValue(10000),
    amtPaidSat: Long.fromValue(10000),
    amtPaidMsat: Long.fromValue(1000000),
    rHash: addInvoiceResponse.rHash,
    // paymentRequest: "payreq",
    expiry: Long.fromValue(40),
    isKeysend: false,
    state: lnrpc.Invoice.InvoiceState.SETTLED,
  })
  act(() => {
    balance = balance - 100;
    DeviceEventEmitter.emit("SubscribeInvoices", {
      data: base64.fromByteArray(lnrpc.Invoice.encode(invoice).finish())
    });
  });

  await wait(async () => {
    expect(store.getState().transaction.transactions[0].status).toBe("SETTLED");
  });

  await wait(() => expect(bigBalanceHeader.children[0]).toContain("0.000999"));
});