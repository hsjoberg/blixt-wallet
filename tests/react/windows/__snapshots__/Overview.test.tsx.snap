// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RNCSafeAreaView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "overflow": "hidden",
          }
        }
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 0.01,
                    "restSpeedThreshold": 0.01,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 0.01,
                    "restSpeedThreshold": 0.01,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                  "marginTop": 64,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    Object {
                      "backgroundColor": "#fff",
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 1,
                        "width": -1,
                      },
                      "shadowOpacity": 0.3,
                      "shadowRadius": 5,
                      "top": 0,
                      "width": 3,
                    }
                  }
                />
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flex": 1,
                            "overflow": "hidden",
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "flex": 1,
                            }
                          }
                        >
                          <View
                            collapsable={false}
                            pointerEvents="auto"
                            removeClippedSubviews={false}
                            style={
                              Array [
                                Object {
                                  "flex": 1,
                                  "overflow": "hidden",
                                },
                                Object {
                                  "bottom": 0,
                                  "left": 0,
                                  "position": "absolute",
                                  "right": 0,
                                  "top": 0,
                                },
                                Object {
                                  "opacity": 1,
                                },
                              ]
                            }
                          >
                            <View
                              style={
                                Object {
                                  "flex": 1,
                                }
                              }
                            >
                              <View
                                accessibilityElementsHidden={false}
                                importantForAccessibility="auto"
                                style={
                                  Array [
                                    Object {
                                      "flex": 1,
                                    },
                                    Object {
                                      "backgroundColor": "rgb(242, 242, 242)",
                                    },
                                  ]
                                }
                              >
                                <View
                                  style={
                                    Array [
                                      Object {
                                        "backgroundColor": "#151314",
                                        "flex": 1,
                                        "height": "100%",
                                        "width": "100%",
                                      },
                                      Object {},
                                      undefined,
                                    ]
                                  }
                                >
                                  <View
                                    style={
                                      Object {
                                        "backgroundColor": "#151314",
                                        "flex": 1,
                                      }
                                    }
                                  >
                                    <ScrollView
                                      contentContainerStyle={
                                        Object {
                                          "paddingBottom": 12,
                                          "paddingLeft": 7,
                                          "paddingRight": 7,
                                          "paddingTop": 207,
                                        }
                                      }
                                      refreshControl={
                                        <RefreshControl
                                          colors={
                                            Array [
                                              "#e5eaea",
                                            ]
                                          }
                                          onRefresh={[Function]}
                                          progressBackgroundColor="#232323"
                                          progressViewOffset={183}
                                          refreshing={false}
                                          title="Refreshing"
                                        />
                                      }
                                      scrollEventThrottle={16}
                                    >
                                      <Text
                                        style={
                                          Array [
                                            Object {
                                              "color": "#000",
                                              "fontFamily": "System",
                                              "fontSize": 16,
                                            },
                                            Object {
                                              "margin": 16,
                                              "textAlign": "center",
                                            },
                                          ]
                                        }
                                        uppercase={false}
                                      >
                                        No transactions yet
                                      </Text>
                                    </ScrollView>
                                    <View
                                      pointerEvents="box-none"
                                      style={
                                        Object {
                                          "alignItems": "center",
                                          "borderBottomColor": "#1F1F1F",
                                          "borderBottomWidth": 1,
                                          "height": 195,
                                          "justifyContent": "center",
                                          "left": 0,
                                          "position": "absolute",
                                          "right": 0,
                                          "top": 0,
                                        }
                                      }
                                    >
                                      <BVLinearGradient
                                        colors={
                                          Array [
                                            4287072135,
                                            4284440415,
                                          ]
                                        }
                                        endPoint={
                                          Object {
                                            "x": 0.5,
                                            "y": 1,
                                          }
                                        }
                                        locations={null}
                                        pointerEvents="box-none"
                                        startPoint={
                                          Object {
                                            "x": 0.5,
                                            "y": 0,
                                          }
                                        }
                                        style={
                                          Object {
                                            "alignItems": "center",
                                            "flex": 1,
                                            "height": "100%",
                                            "justifyContent": "center",
                                            "width": "100%",
                                          }
                                        }
                                      >
                                        <View
                                          style={
                                            Object {
                                              "bottom": 0,
                                              "left": 0,
                                              "position": "absolute",
                                              "right": 0,
                                              "top": 0,
                                            }
                                          }
                                        >
                                          <Text
                                            allowFontScaling={false}
                                            style={
                                              Array [
                                                Object {
                                                  "color": undefined,
                                                  "fontSize": 12,
                                                },
                                                Array [
                                                  Object {
                                                    "color": "#000",
                                                    "fontSize": 30,
                                                  },
                                                  Object {
                                                    "color": "#000",
                                                    "fontSize": 30,
                                                  },
                                                  Object {
                                                    "color": "#e5eaea",
                                                    "fontSize": 25,
                                                    "left": 8,
                                                    "padding": 4,
                                                    "paddingRight": 8,
                                                    "position": "absolute",
                                                    "top": 54,
                                                  },
                                                ],
                                                Object {
                                                  "fontFamily": "FontAwesome",
                                                  "fontStyle": "normal",
                                                  "fontWeight": "normal",
                                                },
                                                Object {},
                                              ]
                                            }
                                            type="FontAwesome"
                                          >
                                            
                                          </Text>
                                          <Text
                                            allowFontScaling={false}
                                            style={
                                              Array [
                                                Object {
                                                  "color": undefined,
                                                  "fontSize": 12,
                                                },
                                                Array [
                                                  Object {
                                                    "color": "#000",
                                                    "fontSize": 30,
                                                  },
                                                  Object {
                                                    "color": "#000",
                                                    "fontSize": 30,
                                                  },
                                                  Object {
                                                    "color": "#e5eaea",
                                                    "fontSize": 28,
                                                    "left": 41,
                                                    "padding": 4,
                                                    "paddingRight": 8,
                                                    "position": "absolute",
                                                    "top": 54,
                                                  },
                                                ],
                                                Object {
                                                  "fontFamily": "Entypo",
                                                  "fontStyle": "normal",
                                                  "fontWeight": "normal",
                                                },
                                                Object {},
                                              ]
                                            }
                                            type="Entypo"
                                          >
                                            
                                          </Text>
                                          <Text
                                            allowFontScaling={false}
                                            style={
                                              Array [
                                                Object {
                                                  "color": undefined,
                                                  "fontSize": 12,
                                                },
                                                Array [
                                                  Object {
                                                    "color": "#000",
                                                    "fontSize": 30,
                                                  },
                                                  Object {
                                                    "color": "#000",
                                                    "fontSize": 30,
                                                  },
                                                  Object {
                                                    "color": "#e5eaea",
                                                    "fontSize": 27,
                                                    "padding": 4,
                                                    "position": "absolute",
                                                    "right": 8,
                                                    "top": 55,
                                                  },
                                                ],
                                                Object {
                                                  "fontFamily": "anticon",
                                                  "fontStyle": "normal",
                                                  "fontWeight": "normal",
                                                },
                                                Object {},
                                              ]
                                            }
                                            type="AntDesign"
                                          >
                                            
                                          </Text>
                                          <Text
                                            allowFontScaling={false}
                                            style={
                                              Array [
                                                Object {
                                                  "color": undefined,
                                                  "fontSize": 12,
                                                },
                                                Array [
                                                  Object {
                                                    "color": "#000",
                                                    "fontSize": 30,
                                                  },
                                                  Object {
                                                    "color": "#000",
                                                    "fontSize": 30,
                                                  },
                                                  Object {
                                                    "color": "#e5eaea",
                                                    "fontSize": 27,
                                                    "padding": 4,
                                                    "position": "absolute",
                                                    "right": 47,
                                                    "top": 55,
                                                  },
                                                ],
                                                Object {
                                                  "fontFamily": "Ionicons",
                                                  "fontStyle": "normal",
                                                  "fontWeight": "normal",
                                                },
                                                Object {},
                                              ]
                                            }
                                          >
                                            
                                          </Text>
                                        </View>
                                        <Text
                                          style={
                                            Object {
                                              "color": "#e5eaea",
                                              "fontFamily": "System",
                                              "fontSize": 37,
                                              "marginBottom": 2,
                                              "marginTop": 72,
                                            }
                                          }
                                        >
                                          0.00000123 ₿
                                        </Text>
                                        <Text
                                          style={
                                            Object {
                                              "color": "#e5eaea",
                                              "fontFamily": "System",
                                              "fontSize": 18,
                                              "opacity": 1,
                                            }
                                          }
                                        >
                                          0.00 USD
                                        </Text>
                                      </BVLinearGradient>
                                    </View>
                                  </View>
                                </View>
                              </View>
                            </View>
                          </View>
                        </View>
                        <View
                          style={
                            Object {
                              "backgroundColor": "#F8F8F8",
                              "borderColor": "#cbcbcb",
                              "borderTopWidth": 0.5,
                              "elevation": 3,
                              "flexDirection": "row",
                              "height": 55,
                              "justifyContent": "center",
                              "left": 0,
                              "paddingBottom": 0,
                              "right": 0,
                            }
                          }
                        >
                          <View
                            style={
                              Object {
                                "alignSelf": "stretch",
                                "backgroundColor": undefined,
                                "flex": 1,
                                "flexDirection": "row",
                                "justifyContent": "space-between",
                              }
                            }
                          >
                            <TouchableOpacity
                              activeOpacity={0.5}
                              style={
                                Object {
                                  "alignItems": "center",
                                  "alignSelf": "center",
                                  "backgroundColor": "transparent",
                                  "borderBottomWidth": null,
                                  "borderColor": null,
                                  "borderLeftWidth": null,
                                  "borderRadius": 5,
                                  "borderRightWidth": null,
                                  "borderTopWidth": null,
                                  "borderWidth": undefined,
                                  "elevation": 0,
                                  "flex": 1,
                                  "flexDirection": null,
                                  "height": 55,
                                  "justifyContent": "center",
                                  "paddingBottom": 6,
                                  "paddingTop": 6,
                                  "shadowColor": null,
                                  "shadowOffset": null,
                                  "shadowOpacity": null,
                                  "shadowRadius": null,
                                }
                              }
                            >
                              <Text
                                allowFontScaling={false}
                                style={
                                  Array [
                                    Object {
                                      "color": undefined,
                                      "fontSize": 12,
                                    },
                                    Array [
                                      Object {
                                        "color": "#000",
                                        "fontSize": 30,
                                      },
                                      Object {
                                        "color": "#6b6b6b",
                                        "fontSize": 24,
                                        "marginLeft": 16,
                                        "marginRight": 16,
                                        "paddingTop": 2,
                                      },
                                    ],
                                    Object {
                                      "fontFamily": "anticon",
                                      "fontStyle": "normal",
                                      "fontWeight": "normal",
                                    },
                                    Object {},
                                  ]
                                }
                                type="AntDesign"
                              >
                                
                              </Text>
                              <Text
                                style={
                                  Object {
                                    "backgroundColor": "transparent",
                                    "color": "#6b6b6b",
                                    "fontFamily": "System",
                                    "fontSize": 14,
                                    "lineHeight": 16,
                                    "marginLeft": 0,
                                    "marginRight": 0,
                                    "paddingLeft": 16,
                                    "paddingRight": 16,
                                  }
                                }
                                uppercase={false}
                              >
                                Receive
                              </Text>
                            </TouchableOpacity>
                          </View>
                          <View
                            style={
                              Object {
                                "alignSelf": "stretch",
                                "backgroundColor": undefined,
                                "flex": 1,
                                "flexDirection": "row",
                                "justifyContent": "space-between",
                              }
                            }
                          >
                            <TouchableOpacity
                              activeOpacity={0.5}
                              style={
                                Object {
                                  "alignItems": "center",
                                  "alignSelf": "center",
                                  "backgroundColor": "transparent",
                                  "borderBottomWidth": null,
                                  "borderColor": null,
                                  "borderLeftWidth": null,
                                  "borderRadius": 5,
                                  "borderRightWidth": null,
                                  "borderTopWidth": null,
                                  "borderWidth": undefined,
                                  "elevation": 0,
                                  "flex": 1,
                                  "flexDirection": null,
                                  "height": 55,
                                  "justifyContent": "center",
                                  "paddingBottom": 6,
                                  "paddingTop": 6,
                                  "shadowColor": null,
                                  "shadowOffset": null,
                                  "shadowOpacity": null,
                                  "shadowRadius": null,
                                }
                              }
                            >
                              <Text
                                allowFontScaling={false}
                                style={
                                  Array [
                                    Object {
                                      "color": undefined,
                                      "fontSize": 12,
                                    },
                                    Array [
                                      Object {
                                        "color": "#000",
                                        "fontSize": 30,
                                      },
                                      Object {
                                        "color": "#6b6b6b",
                                        "fontSize": 24,
                                        "marginLeft": 16,
                                        "marginRight": 16,
                                        "paddingTop": 2,
                                      },
                                    ],
                                    Object {
                                      "fontFamily": "anticon",
                                      "fontStyle": "normal",
                                      "fontWeight": "normal",
                                    },
                                    Object {},
                                  ]
                                }
                                type="AntDesign"
                              >
                                
                              </Text>
                              <Text
                                style={
                                  Object {
                                    "backgroundColor": "transparent",
                                    "color": "#6b6b6b",
                                    "fontFamily": "System",
                                    "fontSize": 14,
                                    "lineHeight": 16,
                                    "marginLeft": 0,
                                    "marginRight": 0,
                                    "paddingLeft": 16,
                                    "paddingRight": 16,
                                  }
                                }
                                uppercase={false}
                              >
                                Send
                              </Text>
                            </TouchableOpacity>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        pointerEvents="box-none"
        style={
          Object {
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          pointerEvents="box-none"
          style={null}
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "rgb(255, 255, 255)",
                  "borderBottomColor": "rgb(199, 199, 204)",
                  "flex": 1,
                  "shadowColor": "rgb(199, 199, 204)",
                  "shadowOffset": Object {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            pointerEvents="box-none"
            style={
              Object {
                "height": 64,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "height": 20,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <View
                pointerEvents="box-none"
                style={
                  Object {
                    "marginHorizontal": 18,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  numberOfLines={1}
                  style={
                    Object {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  a
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RNCSafeAreaView>
  </View>
</View>
`;
